### PR TITLE
Allow arguments for the last flag in short flag batch 

### PR DESCRIPTION
**Pull request by MariaSolOs**
# Description
_Fixes #5923_

Currently `nushell` doesn't allow short flag batches to contain arguments, despite this being a common pattern in commands like `git commit -am 'My commit message'`. This PR relaxes this so that the last flag in the batch can take an argument.

# User-Facing Changes
- `nu::parser::short_flag_arg_cant_take_arg` has been replaced by `nu::parser::only_last_flag_in_batch_can_take_arg` and is displayed when a flag other then the last in a short flag batch takes an argument.

# Tests + Formatting
- Both [`test_parser.rs`](https://github.com/nushell/nushell/blob/48af0ebc3c9f0c7b0ef134078fe1a0e47982059e/crates/nu-parser/tests/test_parser.rs#L640-L704) and [`test_known_external.rs`](https://github.com/nushell/nushell/blob/48af0ebc3c9f0c7b0ef134078fe1a0e47982059e/src/tests/test_known_external.rs#L42-L61) have been updated to test the new allowed and disallowed scenarios.

### CODE REVIEW

**Comment by amtoine** [comment]
:wave: 

quite strangely, it does already work with `git` aliases :open_mouth: 

i have
```bash
alias.cm=commit --verbose
```
and
```bash
git cm -am "foo"
```
works :confused:

**Comment by MariaSolOs** [comment]
> quite strangely, it does already work with `git` aliases 😮
>

@amtoine Very interesting, I have some of the aliases from [`nu_scripts`](https://github.com/nushell/nu_scripts/blob/ebdab3dea34f7776c937497a67357f913b0c1c54/aliases/git/nu_alias_git.nu) and so I didn't consider something like `alias.cm=commit --verbose`.

In fact, could you further explain (or point me to the right docs) what `alias.X` does? I had not seen that before.

**Comment by amtoine** [comment]
> > quite strangely, it does already work with `git` aliases open_mouth
> 
> @amtoine Very interesting, I have some of the aliases from [`nu_scripts`](https://github.com/nushell/nu_scripts/blob/ebdab3dea34f7776c937497a67357f913b0c1c54/aliases/git/nu_alias_git.nu) and so I didn't consider something like `alias.cm=commit --verbose`.
> 
oooh i see what you're using :yum: 
true `nushell` aliases!

> In fact, could you further explain (or point me to the right docs) what `alias.X` does? I had not seen that before.
>
i should have been clearer, sorry :cry: 
when i say i have `alias.cm=commit --verbose`, it means that i've configured `git` to run `commit --verbose` when i simply use the `cm` subcommand! (the `alias.<alias>=...` simply is the output of the `git config --show` command :yum:)
these are built-in `git` aliases, really shortcuts if you will :wink: 

> **Note**
> you can see this one in my [`git` config](https://github.com/goatfiles/dotfiles/blob/nightly/.config/git/config#L44)

**Comment by MariaSolOs** [comment]
> i should have been clearer, sorry 😢
> when i say i have alias.cm=commit --verbose, it means that i've configured git to run commit --verbose when i simply use the cm subcommand! (the alias.<alias>=... simply is the output of the git config --show command 😋)
> these are built-in git aliases, really shortcuts if you will 😉

Oh my no worries at all! I was confused because I thought you were talking about `nushell`'s `alias` hehe. Thanks for clarifying!

File: crates/nu-parser/src/parser.rs [hunk 1]
```diff
@@ -449,24 +449,20 @@ fn parse_short_flags(
...
-        for short_flag in short_flags.iter().enumerate() {
-            let short_flag_char = char::from(*short_flag.1);
+        for (i, short_flag) in short_flags.iter().enumerate() {
+            let short_flag_char = char::from(*short_flag);
             let orig = arg_span;
-            let short_flag_span = Span::new(
-                orig.start + 1 + short_flag.0,
-                orig.start + 1 + short_flag.0 + 1,
-            );
+            let short_flag_span = Span::new(orig.start + 1 + i, orig.start + 1 + i + 1);
```
Comment (sholderbach) [inline crates/nu-parser/src/parser.rs:461]: This line previously already made the assumption that a short flag is only an ASCII character and requires just one byte. Our parsing of definitions apparently doesn't check for that causing downstream panics. I opened an issue for that (#8821)

Comment (MariaSolOs) [reply crates/nu-parser/src/parser.rs:461]: Ohh good catch!

Comment (MariaSolOs) [reply crates/nu-parser/src/parser.rs:461]: @sholderbach would you like #8821 to get fixed first, or are you okay with merging this before?

**Comment by MariaSolOs** [comment]
I've updated the PR to accommodate for the latest changes from #8849 :)

File: crates/nu-parser/src/parser.rs (hunk 1 above)
Comment (MariaSolOs) [reply crates/nu-parser/src/parser.rs]: nvm, it seems like #8849 fixed the issue!

File: crates/nu-parser/src/parser.rs [hunk 2]
```diff
@@ -450,25 +450,22 @@ fn parse_short_flags(
...
                 let orig = arg_span;
-                let short_flag_span = Span::new(
-                    orig.start + 1 + short_flag.0,
-                    orig.start + 1 + short_flag.0 + 1,
-                );
-                if let Some(flag) = sig.get_short_flag(short_flag_char) {
-                    // If we require an arg and are in a batch of short flags, error
-                    if !found_short_flags.is_empty() && flag.arg.is_some() {
-                        working_set.error(ParseError::ShortFlagBatchCantTakeArg(short_flag_span));
-                        break;
```
Comment (sholderbach) [inline crates/nu-parser/src/parser.rs:471]: I think to be defensive we should still break right?

File: crates/nu-parser/src/parser.rs [hunk 3]
```diff
@@ -450,25 +450,22 @@ fn parse_short_flags(
 ) -> Option<Vec<Flag>> {
     let arg_span = spans[*spans_idx];
 
-    let arg_contents = working_set.get_span_contents(arg_span);
+    let arg_contents = working_set.get_span_contents(arg_span).to_vec();
```
Comment (sholderbach) [inline crates/nu-parser/src/parser.rs:453]: We don't need to materialize in this case.

**Review by sholderbach — COMMENTED** [review]
The implementation and your tests look good to me. Thank you very much!

Should be easy to update to the state of #8866

File: crates/nu-parser/src/parser.rs (hunk 2 above)
Comment (MariaSolOs) [reply crates/nu-parser/src/parser.rs:471]: Right. Idk why but I had tried this before and one of my test cases failed with the `break` but it must have been something else.

File: crates/nu-parser/src/parser.rs (hunk 3 above)
Comment (MariaSolOs) [reply crates/nu-parser/src/parser.rs]: Yep, we can remove this if we add the `break`.

**Comment by MariaSolOs** [comment]
@sholderbach For my own learning: Could you please explain what the 88.88% diff coverage refers to in the comment above? I'm reading the [Codecov docs](https://docs.codecov.com/docs/pull-request-comments) and all the other values in the summary table make sense to me, but I'm not understanding where that 88.88% is coming from...

File: crates/nu-parser/src/parser.rs [hunk 4]
```diff
@@ -459,10 +459,9 @@ fn parse_short_flags(
             let mut found_short_flags = vec![];
             let mut unmatched_short_flags = vec![];
             for (offset, short_flag) in short_flags.char_indices() {
-                let orig = arg_span;
```
Comment (MariaSolOs) [inline crates/nu-parser/src/parser.rs:462]: There's no need to keep `orig`.

**Review by amtoine — APPROVED** [review]
looks good to me, that's a sweet improvement :blush:

**Comment by sholderbach** [comment]
> @sholderbach For my own learning: Could you please explain what the 88.88% diff coverage refers to in the comment above? I'm reading the [Codecov docs](https://docs.codecov.com/docs/pull-request-comments) and all the other values in the summary table make sense to me, but I'm not understanding where that 88.88% is coming from...
> 

I think it is the fraction of the added or changed lines that are measured to be covered. But the exact numbers can be weird depending on pure definitions or test modules

**Comment by MariaSolOs** [comment]
> I think it is the fraction of the added or changed lines that are measured to be covered. But the exact numbers can be weird depending on pure definitions or test modules

I still don't get it but all the diffs are green so I'm happy with that 😅

**Comment by sholderbach** [comment]
> I still don't get it but all the diffs are green so I'm happy with that sweat_smile

Me neither :laughing:

**Pull request merged**